python/pyarrow/_gcsfs.pyx

```diff
@@ -0,0 +1,173 @@
+        permissions.
+    credential_token_expiration : datetime, default None
+        Expiration for credential generated with an access token. Must be specified
+        if token is specified.
+    default_bucket_location : str, default 'US-CENTRAL1'
+        GCP region to create buckets in.
+    scheme : str, default 'https'
+        GCS connection transport scheme.
+    endpoint_override : str, default None
+        Override endpoint with a connect string such as "localhost:9000"
+    default_metadata : mapping or pyarrow.KeyValueMetadata, default None
+        Default metadata for open_output_stream.  This will be ignored if
+        non-empty metadata is passed to open_output_stream.
+    """
+
+    cdef:
+        CGcsFileSystem* gcsfs
+
+    def __init__(self, *, bint anonymous=False, access_token=None,
+                 target_service_account=None, credential_token_expiration=None,
+                 default_bucket_location='US-CENTRAL1',
+                 scheme=None,
+                 endpoint_override=None,
+                 default_metadata=None):
+        cdef:
+            CGcsOptions options
+            shared_ptr[CGcsFileSystem] wrapped
+
+        # Intentional use of truthiness because empty strings aren't valid and
+        # for reconstruction from pickling will give empty strings.
+        if anonymous and (target_service_account or access_token):
+            raise ValueError(
+                'anonymous option is not compatible with target_service_account and '
+                'access_token please only specify only one.'
+            )
+        elif ((access_token and credential_token_expiration is None) or
+              (not access_token and
+                  credential_token_expiration is not None)):
+            raise ValueError(
+                'access_token and credential_token_expiration must be '
+                'specified together'
+            )
+
+        elif anonymous:
+            options = CGcsOptions.Anonymous()
+        elif access_token:
+            options = CGcsOptions.FromAccessToken(
+                tobytes(access_token),
+                PyDateTime_to_TimePoint(<PyDateTime_DateTime*>credential_token_expiration))
+        else:
+            options = CGcsOptions.Defaults()
+
+        if target_service_account:
+            options = CGcsOptions.FromImpersonatedServiceAccount(
+                options.credentials, tobytes(target_service_account))
+
+        options.default_bucket_location = tobytes(default_bucket_location)
+
+        if scheme is not None:
+            options.scheme = tobytes(scheme)
+        if endpoint_override is not None:
+            options.endpoint_override = tobytes(endpoint_override)
+        if default_metadata is not None:
+            if not isinstance(default_metadata, KeyValueMetadata):
+                default_metadata = KeyValueMetadata(default_metadata)
+            options.default_metadata = pyarrow_unwrap_metadata(
+                default_metadata)
```

Comment:
It should be possible to write this:
```suggestion
            options.default_metadata = pyarrow_unwrap_metadata(
                ensure_metadata(default_metadata))
```

Perhaps you'll also need to declare `ensure_metadata` in `lib.pxd`:
```diff
diff --git a/python/pyarrow/lib.pxd b/python/pyarrow/lib.pxd
index 5b9d3001b..b2aa69afa 100644
--- a/python/pyarrow/lib.pxd
+++ b/python/pyarrow/lib.pxd
@@ -531,6 +531,7 @@ cdef shared_ptr[CInputStream] native_transcoding_input_stream(
 
 # Default is allow_none=False
 cpdef DataType ensure_type(object type, bint allow_none=*)
+cdef KeyValueMetadata ensure_metadata(object meta, c_bool allow_none=*)
 
 cdef timeunit_to_string(TimeUnit unit)
 cdef TimeUnit string_to_timeunit(unit) except *
```